cmd/modelcmd/controller.go

```diff
@@ -75,28 +88,73 @@ func (c *ControllerCommandBase) SetClientStore(store jujuclient.ClientStore) {
 
 // ClientStore implements the ControllerCommand interface.
 func (c *ControllerCommandBase) ClientStore() jujuclient.ClientStore {
+	c.assertRunStarted()
 	return c.store
 }
 
-// SetControllerName implements ControllerCommand.SetControllerName.
-func (c *ControllerCommandBase) SetControllerName(controllerName string, allowDefault bool) error {
+func (c *ControllerCommandBase) initController() error {
```

Comment:
This initController/initModel business feels pretty gross, but I guess it's all we can do for now.

I think we eventually need to move away from the inheritance pattern we currently have around commands, and towards composition so that the ModelCommandBase (or whatever it's called now) Run method comes first, and delegates to the command-specific code with a separate signature, like:

```go
package modelcmd

type ModelCommand interface {
    Init(...)
    SetFlags(...)
    Run(modelcmd.ModelContext) error
}

type ControllerContext struct {
    Store jujuclient.ClientStore
    APIDialer APIDialer
    ControllerName string
}

type ModelContext struct {
    ControllerContext
    ModelName string
}

type APIDialer interface {
    DialAPI(controllerName, modelName string) (api.Connection, error)
}
```

(implementations of ModelCommand wouldn't embed anything)